nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+        } else {
+            let total_observe = u128::try_from(self.observations[last_index].num_of_observations).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].num_of_observations).unwrap();
+            let price1cumu = u128::try_from(self.observations[last_index].price1_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price1_cumulative).unwrap();
+            let price2cumu = u128::try_from(self.observations[last_index].price2_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price2_cumulative).unwrap();
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (U128(mean1), U128(mean2));
+        }
+    }
+
+    pub(crate) fn log_observation(&mut self, pool: PoolInfo) {
+        // price1, price2 calculate
+        let near = u128::try_from(pool.ynear).unwrap();
+        let reserve = u128::try_from(pool.reserve).unwrap();
+        let price1: u128 = near / reserve;
+        let price2: u128 = reserve / near;
+
+        // update current index
+        if self.populated == 0 {
+            self.last_updated_index = self.initialize(env::block_timestamp(), price1, price2);
+            self.populated += 1;
+        } else {
+            self.last_updated_index = self.write(
+                env::block_timestamp(),
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        let len: usize;
+        if self.pivoted == true {
+            len = MAX_LENGTH.into();
+        } else {
+            len = self.observations.len();
+        }
+
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct Observation {
+    // timestamp
+    pub block_timestamp: U64,
+    // Number of observations till block_timestamp
+    pub num_of_observations: U128,
+    // cumulative price of token1 till block_timestamp
+    pub price1_cumulative: U128,
+    // cumulative price of token2 till block_timestamp
+    pub price2_cumulative: U128,
+}
+
+impl Observation {
+    /**
+    Transforms a previous observation into a new observation.
+    Parameters:
+    + `block_timestamp`: _must_ be chronologically equal to or greater than last.block_timestamp.
+    + `last`: The specified observation to be transformed.
+    + `price1`: price of first token.
+    + `price2`: price of second token.
+    */
+    pub fn transform(
+        last: &Observation,
+        block_timestamp: u64,
+        price1: u128,
+        price2: u128
+    ) -> Observation {
```

Comment:
This can be a method:
```
pub fn transform(&self, block_timestamp...)
```

or... why not mutating?
```
pub fn update(&mut self, block_timestamp: u64 ...)
```

Reply:
not mutating because this gives transformed value from last observed value,
it also considers same timestamp readings: it increases num_of_observations by 1 for current reading
therefore it seemed more clear.
otherwise it could be added in write method.

Reply:
so, if you don't want to mutate, then why not doing a method which will return a new Observation (as in the first suggestion)?

Reply:
a method which is out of class ?

Reply:
`Observation`

Reply:
ok